File: JUNE 23/01 June/Topological sort gfg.cpp
class Solution
{
	public:
	//Function to return list containing vertices in Topological order. 
	vector<int> topoSort(int V, vector<int> adj[]) 
	{
	    // code here
	    vector<int> degree(V+5,0);
	    vector<bool> visited(V+5,false);

	    for(int i=0; i<V; i++)
	    {
	        for(int node = 0; node<adj[i].size(); node++)
	        {
	            degree[adj[i][node]]++;
	        }
	    }
	   // for(int i=0; i<V; i++)
	   //     cout<<degree[i]<<" ";
	   // cout<<endl;
	    queue<int> q;
	    for(int i=0;i<V;i++)
	    {
	        if(degree[i]==0)
	        {
	            q.push(i);
	        }
	    }
	    vector<int> topo;
	    while(!q.empty())
	    {
	        int node = q.front();
	        topo.push_back(node);
	        q.pop();
	        for(int i=0; i<adj[node].size(); i++)
	        {
	            degree[adj[node][i]]--;
	            if(degree[adj[node][i]]==0)
	            {
	                q.push(adj[node][i]);
	            }
	        }
	    }
	   // for(int i=0; i<V; i++)
	   //     cout<<topo[i]<<" ";
	   // cout<<endl;
	    return topo;
	}
};

File: APRIL 2025/06 April/Topological sort gfg.java
class Solution {
    public static ArrayList<Integer> topoSort(int V, int[][] edges) {
        // code here
        int[] indegree = new int[V];
        
        List<List<Integer>> adj = new ArrayList<>(V);
        for(int i=0; i<V; i++)
            adj.add(new ArrayList<>());
            
        int EdgeNo = edges.length;
        for(int i=0; i<EdgeNo; i++)
        {
            adj.get(edges[i][0]).add(edges[i][1]);
            indegree[edges[i][1]]++;
        }
        
        Queue<Integer> q = new ArrayDeque<>();
        for(int i=0; i<V; i++) // Initialize queue with the vertices who has no incoming edge
        {
            if(indegree[i] == 0)
                q.offer(i);
        }
        
        ArrayList<Integer> topoTraversal = new ArrayList<>();
        while(!q.isEmpty())
        {
            int node = q.poll();
            topoTraversal.add(node);
            for(Integer next:adj.get(node))    
            {
                indegree[next]--;
                if(indegree[next]==0)
                    q.offer(next);
            }
        }
        return topoTraversal;
    }
}
